backfill/2022-01-13-regrets-reporter-ucs/README.md

```diff
@@ -0,0 +1,101 @@
+
+```bash
+cd path/to/gcp-ingestion
+
+./path/to/bigquery-backfill/backfill/2022-01-13-regrets-reporter-ucs/02-gcp-ingestion-beam-decoder.sh
+```
+
+That decoder job is configured to populate the tables in the `moz-fx-data-backfill-10.regrets_reporter_ucs_live` dataset and write any errors to the `moz-fx-data-backfill-10.payload_bytes_error.backfill` table.
+
+Output:
+  - 822,830 `moz-fx-data-backfill-10.regrets_reporter_ucs_live.main_events_v1` rows.
+  - 2 `moz-fx-data-backfill-10.regrets_reporter_ucs_live.video_index_v1` rows.
+
+
+## Step 3:  Copy & deduplicate decoded RegretsReporter events
+
+To copy and deduplicate the rows from the "live" tables in the `moz-fx-data-backfill-10.regrets_reporter_ucs_live` dataset into the "stable" tables in the `moz-fx-data-backfill-10.regrets_reporter_ucs_stable` dataset, run the [`03-bigquery-etl-copy-deduplicate.sh` script](03-bigquery-etl-copy-deduplicate.sh):
+
+```bash
+cd path/to/bigquery-etl
+
+./path/to/bigquery-backfill/backfill/2022-01-13-regrets-reporter-ucs/03-bigquery-etl-copy-deduplicate.sh
+```
+
+Output:
+  - 822,368 `moz-fx-data-backfill-10.regrets_reporter_ucs_stable.main_events_v1` rows.
+  - 2 `moz-fx-data-backfill-10.regrets_reporter_ucs_stable.video_index_v1` rows.
+
+
+## Step 4:  Insert previously rejected RegretsReporter events into production tables
+
+To insert the previously rejected RegretsReporter events into production tables, someone with appropriate access can run the following `bq` commands either in Google Cloud Shell or using the Google Cloud SDK:
+
+```bash
+bq cp --append_table moz-fx-data-backfill-10:regrets_reporter_ucs_live.main_events_v1 moz-fx-data-shared-prod:regrets_reporter_ucs_live.main_events_v1
+
+bq cp --append_table moz-fx-data-backfill-10:regrets_reporter_ucs_live.video_index_v1 moz-fx-data-shared-prod:regrets_reporter_ucs_live.video_index_v1
+
+bq cp --append_table moz-fx-data-backfill-10:regrets_reporter_ucs_stable.main_events_v1 moz-fx-data-shared-prod:regrets_reporter_ucs_stable.main_events_v1
+
+bq cp --append_table moz-fx-data-backfill-10:regrets_reporter_ucs_stable.video_index_v1 moz-fx-data-shared-prod:regrets_reporter_ucs_stable.video_index_v1
+```
+
+Output:
+  - 822,830 `moz-fx-data-shared-prod.regrets_reporter_ucs_live.main_events_v1` rows.
+  - 2 `moz-fx-data-shared-prod.regrets_reporter_ucs_live.video_index_v1` rows.
+  - 822,368 `moz-fx-data-shared-prod.regrets_reporter_ucs_stable.main_events_v1` rows.
+  - 2 `moz-fx-data-shared-prod.regrets_reporter_ucs_stable.video_index_v1` rows.
+
+
+## Step 5:  Clean up
+
+Data SRE will remove the [resources that were created for this backfill with Terraform](https://github.com/mozilla-services/cloudops-infra/pull/3694) using Terraform.
+
+Any GCS buckets created by Dataflow jobs can be removed like so:
+
+```bash
+# BE CAREFUL!  This removes all GCS buckets from the target project.  It cannot be undone.
+
+gsutil ls -p moz-fx-data-backfill-10 | xargs gsutil -m rm -r
+```
```

Comment:
Deleting from `payload_bytes_error` seems odd to me, as it erases some history of the error having occurred and the newly ingested rows aren't making it into `payload_bytes_decoded` (should they?).

I presume there's a good reason for doing that though, so I've gone ahead and pushed a commit with the DELETE query.  r?

Reply:
> Deleting from payload_bytes_error seems odd to me, as it erases some history of the error having occurred

This is a good question. I'd argue the purpose of payload_bytes_error is primarily to enable backfills like this when incorrect processing has occurred. It also serves as a source for monitoring error cases. This backfill is about recovering from an error case that should no longer exist (since the source bug has been fixed), so I don't think we get much benefit from having a history of the error occurring. We also would like to set a shorter retention period on errors; we should expect these to eventually expire.

The primary benefit of deleting from the error table is ensuring we don't backfill them again. I think a given row should appear _either_ in errors _or_ in stable tables. The count of distinct doc_ids in `raw` should equal the count of distinct doc_ids in `error` + the count of distinct doc_ids in `stable`.

> the newly ingested rows aren't making it into payload_bytes_decoded (should they?).

`decoded` exists largely for debugging or backfill scenarios. It has a short retention period (14 days IIRC) so the relevant records would mostly have rolled off by now.

cc @whd for thoughts on either of these points.

Reply:
> The primary benefit of deleting from the error table is ensuring we don't backfill them again. I think a given row should appear either in errors or in stable tables. The count of distinct doc_ids in raw should equal the count of distinct doc_ids in error + the count of distinct doc_ids in stable.

I generally agree with this, but want to note that we've probably been inconsistent about this in the past.

> decoded exists largely for debugging or backfill scenarios. It has a short retention period (14 days IIRC) so the relevant records would mostly have rolled off by now.

I agree pbd doesn't need to be backfilled, as they have indeed rolled off.